Sources/SwiftSyntax/SyntaxText.swift

```diff
@@ -0,0 +1,226 @@
+  }
+
+  /// Base address of the memory range this string refers to.
+  public var baseAddress: UnsafePointer<UInt8>? {
+    buffer.baseAddress
+  }
+
+  /// Byte length of this string.
+  public var count: Int {
+    buffer.count
+  }
+
+  /// A Boolean value indicating whether a string has no characters.
+  public var isEmpty: Bool {
+    buffer.isEmpty
+  }
+
+  /// Returns `true` if the memory range of this string is a part of `other`.
+  ///
+  /// `text[n ..< m].isSliceOf(text)` is always true as long as `n` and `m` are
+  /// valid indices.
+  public func isSliceOf(_ other: SyntaxText) -> Bool {
+    guard !isEmpty && !other.isEmpty else {
+      return isEmpty == other.isEmpty
+    }
+    return (other.baseAddress! <= baseAddress! &&
+            baseAddress! + count <= other.baseAddress! + other.count)
+  }
+
+  /// Returns `true` if `other` is a substring of this `SyntaxText`.
+  ///
+  /// This should not be used because `SyntaxText` is a `Collection<UInt8>` and
+  /// `contains(_:)` is usually for checking a collection contains the element.
+  /// Using `contains(_:)` for subsequence is confusing.
+  @available(*, deprecated, message: "Use 'firstRange(of:) != nil'")
+  public func contains(_ other: SyntaxText) -> Bool {
+    return firstRange(of: other) != nil
+  }
+
+  /// Finds and returns the range of the first occurrence of `other` within this
+  /// string. Returns `nil` if `other` is not found.
+  public func firstRange(of other: SyntaxText) -> Range<Index>? {
+    if other.isEmpty { return nil }
+    let needle = other.baseAddress!
+    let stop = self.count - other.count
+    var start = 0
+    // If 'other' is longer than 'self', stop < 0.
+    while start <= stop {
+      if compareMemory(self.baseAddress! + start, needle, other.count) {
+        return start ..< (start + other.count)
+      } else {
+        start += 1
+      }
+    }
+    return nil
+  }
+
+  /// Returns `true` if the string begins with the specified prefix.
+  public func hasPrefix(_ other: SyntaxText) -> Bool {
+    guard self.count >= other.count else { return false }
+    guard !other.isEmpty else { return true }
+    return self[0..<other.count] == other
+  }
+
+  /// Returns `true` if the string ends with the specified suffix.
+  public func hasSuffix(_ other: SyntaxText) -> Bool {
+    guard self.count >= other.count else { return false }
+    guard !other.isEmpty else { return true }
+    return self[(self.count - other.count)..<self.count] == other
+  }
+}
+
+/// `SyntaxText` is a collection of `UInt8`.
+extension SyntaxText: RandomAccessCollection {
+  public typealias Element = UInt8
+  public typealias Index = Int
+  public typealias SubSequence = SyntaxText
+
+  public var startIndex: Index { buffer.startIndex }
+  public var endIndex: Index { buffer.endIndex }
+
+  public subscript(bounds: Range<Index>) -> SyntaxText {
+    if isEmpty && bounds.isEmpty { return self }
```

Comment:
Do you think it would be worth adding an 
```swift
assert(bounds.upperBound < count) 
```
here to catch any memory issues?

Reply:
Ah, sure.

Reply:
I removed `subscript(bounds: Range<Index>)` because the default implementation is provided as I changed the `Subsequence` to `Slice<Self>`